### ISSUE TITLE
KEGG REST API return types in Python 3

### ISSUE THREAD

**Issue by widdowquinn** [biopython/biopython#832]
Python 3 makes a distinction between string types (`String` and `Byte`) that was not made in Python 2:

```
Python 2.7.11 (default, Mar 22 2016, 08:29:04) 
[GCC 4.2.1 Compatible Apple LLVM 7.3.0 (clang-703.0.29)] on darwin
Type "help", "copyright", "credits" or "license" for more information.
>>> 'A' == b'A'
True
```

```
Python 3.5.1 (default, Dec 26 2015, 18:11:22) 
[GCC 4.2.1 Compatible Apple LLVM 7.0.2 (clang-700.1.81)] on darwin
Type "help", "copyright", "credits" or "license" for more information.
>>> 'A' == b'A'
False
```

The `Bio.KEGG.REST` functions return Byte types, rather than String types:

```
>>> from Bio.KEGG.REST import *
>>> type(kegg_info("kegg").read())
<class 'bytes'>
>>> kegg_info("eco").read()
b'T00007           Escherichia coli K-12 MG1655 KEGG Genes Database\neco              Release 78.0+/05-21, May 16\n                 Kanehisa Laboratories\n                 4,497 entries\n'
```

These give unintuitive results with `print()`:

```
>>> print(kegg_info("eco").read())
b'T00007           Escherichia coli K-12 MG1655 KEGG Genes Database\neco              Release 78.0+/05-21, May 16\n                 Kanehisa Laboratories\n                 4,497 entries\n'
```

and when attempting to parse sequence data returned from KEGG with `Bio.SeqIO`:

```
>>> from Bio import SeqIO
>>> print(kegg_get("ece:Z5100", "ntseq").read())
b'>ece:Z5100 espF; espF; K12786 LEE-encoded effector EspF (N)\natgcttaatggaattagtaacgctgcttctacactagggcggcagcttgtaggtatcgca\nagtcgagtgagctctgcggggggaactggattttctgtagcccctcaggccgtgcgtctt\nactccggtgaaagttcattcccctttttctccaggctcgtcgaatgttaatgcgagaacg\natttttaatgtgagcagccaggtgacttcatttactccctctcgtccggcaccgccgcca\nccgacaagtggacaggcatccggggcatcccgacctttaccgcccattgcacaggcatta\naaagagcacttggctgcctatgaaaaatcgaaaggtcctgaggctttaggttttaagccc\ngcccgtcaggcaccgccgccaccgacaagtggacaggcatccggggcatcccgaccttta\nccgcccattgcacaggcattaaaagagcacttggctgcctatgaaaaatcgaaaggtcct\ngaggctttaggttttaagcccgcccgtcaggcaccgccgccaccgacaagtggacaggca\ntccggggcatcccgacctttaccgcccattgcacaggcattaaaagagcacttggctgcc\ntatgaaaaatcgaaaggtcctgaggctttaggttttaagcccgcccgtcaggcaccaccg\nccaccgacagggcctagtggactaccgccccttgcacaggcattaaaagatcatttagct\ngcctatgagcaatcgaagaaagggtaa\n'
>>> SeqIO.read(kegg_get("ece:Z5100", "ntseq"), 'fasta')
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/usr/local/lib/python3.5/site-packages/Bio/SeqIO/__init__.py", line 666, in read
    first = next(iterator)
  File "/usr/local/lib/python3.5/site-packages/Bio/SeqIO/__init__.py", line 602, in parse
    for r in i:
  File "/usr/local/lib/python3.5/site-packages/Bio/SeqIO/FastaIO.py", line 124, in FastaIterator
    for title, sequence in SimpleFastaParser(handle):
  File "/usr/local/lib/python3.5/site-packages/Bio/SeqIO/FastaIO.py", line 48, in SimpleFastaParser
    if line[0] == ">":
IndexError: index out of range
```

Under Python 2, these returned results were silently treated as String types, so `print()` and `SeqIO.read()` worked as would intuitively be expected.

We could modify `Bio.KEGG.REST` to return String, rather than Byte streams, but this would cause problems for requests for image objects, which are returned as GIF format, e.g.

```
>>> print(kegg_get("cpd:C01290", "image").read())
b'GIF87a\x10\x02\xc6\x00\xf7\x00\x00\x00\x00\x00\xb1\xb1\xb1\xaf\xaf\xaf\xad\xad\xad\xab\xab\xab\xa9\xa9\xa9\xa7\xa7\xa7\[…]
```

I would suggest that, to restore expected behaviour to be comparable to Python 2, we modify `kegg_get()` to return StringIO stream versions of decoded Byte streams for the KEGG text responses, and the native response Byte stream for image responses.

I'm happy to take a pass at this, but I am wondering if there is a better/preferable approach we might try?

**Comment by peterjc** [comment]
We made a similar pragmatic choice in the ``Bio.Entrez`` wrapper to wrap the bytes-handles to make them into plain text-handles via:

``` python
from Bio._py3k import _binary_to_string_handle
```

This exposed further issues with XML parsers (some actually like binary mode and bytes), but that is now all working.

Again, the logic here was that (almost?) all the datatypes Entrez returned were text based (including XML), and almost all our parsers expected text based handles rather than binary bytes handles.

So I'm OK with your proposal - leave image handles as raw bytes, but for everything else return ``_binary_to_string_handle(handle)`` instead?

**Comment by widdowquinn** [comment]
Thanks Peter - that sounds good.

I'll use the `Bio.Entrez` code as a guide, and you can expect a pull request later.

**Comment by peterjc** [comment]
This was fixed with #833, thanks!
